pycsw/wsgi_flask.py

```diff
@@ -30,29 +30,33 @@
 # =================================================================
 
 import os
+from configparser import ConfigParser
+from pathlib import Path
 
 from flask import Flask, Blueprint, make_response, request
 
+from pycsw.core.util import EnvInterpolation
 from pycsw.ogc.api.records import API
 from pycsw.ogc.api.util import STATIC
 from pycsw.wsgi import application_dispatcher
 
-PYCSW_CONFIG = os.environ.get('PYCSW_CONFIG')
 
-if PYCSW_CONFIG is None:
-    raise RuntimeError('PYCSW_CONFIG environment variable not set')
+def _parse_config(config_path: Path) -> ConfigParser:
+    config = ConfigParser(interpolation=EnvInterpolation())
+    with config_path.open(encoding='utf-8') as scp:
+        config.read_file(scp)
+    return config
 
-api_ = API(PYCSW_CONFIG)
```

Comment:
I see that the `API` object is now instantiated within each route. What's the benefit/rationale?

Reply:
The idea is to be able to import the `wsgi_flask` module with little side-effects, to make it easier to test pycsw. Testing makes use of a `Flask.test_client` structure, as mentioned in the [flask docs](https://flask.palletsprojects.com/en/2.0.x/testing/#the-testing-skeleton). And in order to use this `test_client` we need to be able to import the module that defines the flask application, which in this case is `pycsw.wsgi_flask`.

Previously, `API` was a global variable that was being accessed inside each route. Since it takes the pycsw configuration as an initialization parameter, this meant that there was not a clean way to provide a custom config for pycsw. By the time we created the flask `test_client` the pycsw config was already read and the `API` was already instantiated. This made it cumbersome to write the tests, as we would have to resort to some mocking, which IMO seemed like unnecessary complexity.

With this change, it becomes possible to modify pycsw config at any time, by interacting with the flask config object. The fact that API is instantiated inside each route means we can provide a custom configs in the tests.

Let me know if you feel strongly against this and I can try to come up with some alternative solution (most likely using mocks for testing).

Reply:
Is the approach taken in the [pygeoapi](https://github.com/geopython/pygeoapi/blob/master/tests/test_api.py) API tests something we should consider?  In other words, we test the `pycsw.ogc.api.records` itself / directly?

Reply:
I would add both.

Testing the various functions directly (AKA unit testing) is nice to catch logic errors. Unit tests verify implementation details, so they usually require that the person writing the test to know the underlying code well.


Testing how it all interacts together (AKA functional testing) is nice to check the big picture of things. Functional tests focus on correctness of outputs, so they are usually more of a black-box type of thing. The person writing the test just needs to check if the expected result is met.

Both of these are useful. 

Usually the OGC test suites focus on functional testing, as that is what gives you some assurance about the overall behavior of the system being as expected.

In our older code we actually have both approaches too. If you look at `pycsw/tests` there is a `functionaltests` (where the test suites are) and a `unittests` directories.

Reply:
Reading the config on every api call adds an overhead though, correct?